updates for crypto scalar instructions

anku-anand opened the pull request.

anku-anand committed:
1)  pack and zext.h have same opcode, func3, funct7 only diffrence is in rs2 value
 for zext.h rs2 is always 0, if pack instruction is used with x0 as rs2
 then cannot distinguish from each other, hence using cover label to differentiate.

2) updated coverage.py for xperm4, xperm8, zip, unzip instructions

anku-anand committed:
Merge branch 'master' of https://github.com/anku-anand/riscv-isac

pawks commented on riscv_isac/plugins/internaldecoder.py:
```diff
@@ -1210,7 +1211,10 @@ def arith_ops(self, instrObj):
+# pack and zext.h have same opcode, func3, funct7 only diffrence is in rs2 value
+# for zext.h rs2 is always 0, if pack instruction is used with x0 as rs2
+# then cannot distinguish from each other, hence using cover label to differentiate.
+                if rs2[0] == 0b0 and  "pack" not in self.labels:
```
This is not the right way to deal with pseudo ops. This will cause problems when the input labels contain both `pack` and `zext.h`. Pseudo-ops need some further thought. There are 2 ways to handle this. The first is to statically encode the pseudo-ops in the decoder and add another field in the instruction object which contains the list of pseudo ops satisfied by that instruction instance. The second is to express the coverpoints for the pseudo ops as the original ops and add another node in the `covergroup` node called `p_op_cond:`. So the condition to check whether a particular instruction should be considered for a covergroup also checks if the `p_op_cond` is satisfied before updating the coverpoints. In this case `zext.h` covergroup becomes:
```
zext.h:
    config: 
      - check ISA:=regex(.*I.*B.*)
      - check ISA:=regex(.*I.*Zbb.*)
    opcode: 
      pack: 0
      packw: 0
    p_op_cond: rs2 == x0
    rs1: 
      <<: *all_regs
    rd: 
      <<: *all_regs
    op_comb: 
      <<: *ifmt_op_comb
    val_comb:
      'rs1_val == 0': 0
      'rs1_val == 0x80': 0
      'rs1_val == 0xFF80': 0
      abstract_comb:
        'walking_ones("rs1_val", xlen, False)': 0
        'walking_zeros("rs1_val", xlen, False)': 0
        'uniform_random(20, 100, ["rs1_val"], [xlen])': 0
``` 

Thoughts? @allenjbaum @neelgala

pawks commented on riscv_isac/plugins/internaldecoder.py:
```diff
@@ -406,8 +406,9 @@ def init_rvp_dictionary(self):
 
     @plugins.decoderHookImpl
-    def setup(self, arch):
+    def setup(self, arch,labels):
```
The decoder should not be dependent on the `labels` to decode instructions.

anku-anand committed:
using isa to differentiate between zext.h and pack, packw instructions

anku-anand commented:
In the Bitmanip or crypto scalar spec, zext.h is not mentioned as pseudo instruction.
Only zext.w is mentioned as pseudo instruction for add.uw, zext.w is not included in the spec.
zext.h is only in zbb spec. pack and packw are only in zbkb. In the instruction decoder we
can use the isa string to differentiate between zext.h  and packw or pack with rs2 as x0.

pawks commented:
> In the Bitmanip or crypto scalar spec, zext.h is not mentioned as pseudo instruction. Only zext.w is mentioned as pseudo instruction for add.uw, zext.w is not included in the spec. zext.h is only in zbb spec. pack and packw are only in zbkb. In the instruction decoder we can use the isa string to differentiate between zext.h and packw or pack with rs2 as x0.

This causes the same problem as before. What if both the nodes are present in the cgf? Then the decoder will always give out the instruction as `zext.h`.

anku-anand commented:
Hi,

Even if both the instructions are in the same cgf, the config part of the
instructions are different. zext will have zbb and pack, packw will have
zbkb.
Since zext and pack will be in different isa spec, ISA strings mentioned in
the config can be used to differentiate.

Also, so far for each spec we have a separate signature file. I am not sure
if we will have zbb and zbkb in the same cgf.

Thanks & Regards,
Anand

On Tue, Apr 12, 2022 at 11:15 AM S Pawan Kumar ***@***.***>
wrote:

> In the Bitmanip or crypto scalar spec, zext.h is not mentioned as pseudo
> instruction. Only zext.w is mentioned as pseudo instruction for add.uw,
> zext.w is not included in the spec. zext.h is only in zbb spec. pack and
> packw are only in zbkb. In the instruction decoder we can use the isa
> string to differentiate between zext.h and packw or pack with rs2 as x0.
>
> This causes the same problem as before. What if both the nodes are present
> in the cgf? Then the decoder will always give out the instruction as
> zext.h.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv-software-src/riscv-isac/pull/34#issuecomment-1096100185>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AECK3Y2ATGWYNK24EGJI6S3VEUEWXANCNFSM5R2YZ6PA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

anku-anand committed:
Merge branch 'riscv-software-src:master' into master

anku-anand committed:
Added gorci to unsgn_rs1 in coverage.py
Added conditions to extract bs and rnum fields in rvopcodesdecoder.py
Removed zext.h from internaldecoder.py

anku-anand commented:
Updated cgf  to use base_op and p_op_cond for pseudo instructions. Removed the condition to check for labels.

pawks commented on riscv_isac/plugins/specification.py:
```diff
@@ -5,7 +5,7 @@
 class DecoderSpec(object):
     @decoderHookSpec
-    def setup(self,arch):
+    def setup(self,arch,isa):
```
The setup function of the plugins does not need the `isa` anymore and all of the changes related need to be reverted back.

anku-anand replied:
Removed isa from the plugin setup function

anku-anand committed:
removed isa argument from decoder plugin setup

anku-anand committed:
Merge branch 'master' into master

anku-anand committed:
update CHANGELOG.md

jamesbeyond commented:
Dear @anku-anand 

The CTG and ISAC Repo had been moved to [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) via PR
https://github.com/riscv-non-isa/riscv-arch-test/pull/495

If you think this PR is important, and need to me merged on to risc-v CTG or ISAC, please re-submit your PR against   [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test)

This repository shall be archived in about one week time .
